### ISSUE TITLE
Duplicate History when I click on a gatsby link

### ISSUE THREAD

**Issue by IgordeOliveira** [gatsbyjs/gatsby#25424]
## Description

I'm creating a multi-language site, so what I did was delete the original pages, and recreate them with the language slug in front resulting in: /pages/page1 => /en/page1, all this part of multilanguage works perfectly, but a problem arose at least strange, when I click on a link gatsby within the site, and I go to an internal page for example, it duplicates that page in the browser history, this causes me to have to click 2 times on the button back to be able to return to the previous page.

### Steps to reproduce

Go to https://duplicate-history.underlab.com.br/en
Go to Products page ( click on products in navbar )
Try to go back to the page by clicking the back button in the browser

### Expected result

Back to /en page
### Actual result

remains on the same page, changing only the title

### Environment

  System:
    OS: Windows 10 10.0.19041
    CPU: (4) x64 Intel(R) Core(TM) i5-4460  CPU @ 3.20GHz
  Binaries:
    Node: 12.16.3 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.4 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 6.14.5 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.8.2 - /c/Python38/python
  Browsers:
    Edge: 44.19041.1.0
  npmPackages:
    gatsby: ^2.23.12 => 2.23.12 
    gatsby-cli: ^2.12.51 => 2.12.51 
    gatsby-image: ^2.4.9 => 2.4.9 
    gatsby-plugin-google-tagmanager: ^2.3.7 => 2.3.7 
    gatsby-plugin-manifest: ^2.4.14 => 2.4.14 
    gatsby-plugin-prefetch-google-fonts: ^1.4.3 => 1.4.3 
    gatsby-plugin-react-helmet: ^3.3.6 => 3.3.6 
    gatsby-plugin-sharp: ^2.6.14 => 2.6.14 
    gatsby-plugin-sitemap: ^2.4.7 => 2.4.7 
    gatsby-plugin-styled-components: ^3.3.6 => 3.3.6 
    gatsby-source-filesystem: ^2.3.14 => 2.3.14 
    gatsby-transformer-sharp: ^2.5.7 => 2.5.7 
  npmGlobalPackages:
    gatsby: 2.21.10

**Comment by graysonhicks** [comment]
Just followed reproduction steps and cannot reproduce.  I get the expected behavior returning to `/en`.

**Comment by vladar** [comment]
Also could not reproduce this. Worked fine for me.

**Comment by IgordeOliveira** [comment]
Oops, I'm sorry, I ended up doing a "workaround" yesterday and I did deploy it, I climbed the old version at this address: https://duplicate-history.underlab.com.br/en
What I did that worked ( but do not know why ) was:

gatsby-browser.js
```
const { navigate } = require("gatsby");

exports.onPreRouteUpdate = ({ location, prevLocation }) => {
  if (prevLocation?.pathname !== location.pathname) { // BUGFIX: https://github.com/gatsbyjs/gatsby/issues/25424
    navigate(location.pathname, {
      replace: true
    })
  }
}
```
@graysonhicks Yes I always use gatsby clean before doing build, and my link is exactly the same as gatsby documentation, I import { Link } from 'gatsby' and use <Link to={href} >Label</Link>

**Comment by vladar** [comment]
Somehow when you go from `/en` to `/en/products` two `/en/products` entries are added to the history stack. So when you call back you still end up on the same URL.

Can you share your code or prepare a [minimal reproduction][reproduction]? If you're up for it, we'd very much appreciate if you could [provide a minimal reproduction][reproduction] and we'll be able to take another look.

Thanks for using Gatsby! 💜

[reproduction]: https://gatsby.dev/reproduction

**Comment by IgordeOliveira** [comment]
Hi @vladar, I tried in several ways to create a minimal reproduction, but when I put only my language logic ( of gatsby-node ) in a new project of gatsby works normally, my last attempt is to add you in the repository, I hope you can help me identify what causes it
https://github.com/IgordeOliveira/gatsby-duplicate-history

**Comment by blainekasten** [comment]
@IgordeOliveira what does `yarn why @reach/router` print out? I'm thinking you might be on an old version of the router that had this bug.

**Comment by IgordeOliveira** [comment]
@blainekasten 
```
yarn why @reach/router
yarn why v1.22.4
[1/4] Why do we have the module "@reach/router"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@reach/router@1.3.4"
info Reasons this module exists
   - "gatsby" depends on it
   - Hoisted from "gatsby#@reach#router"
info Disk size without dependencies: "228KB"
info Disk size with unique dependencies: "472KB"
info Disk size with transitive dependencies: "632KB"
info Number of shared dependencies: 9
Done in 1.62s.

```

**Comment by blainekasten** [comment]
Yup, this is definitely cause of an old version of reach-router. 

![image](https://user-images.githubusercontent.com/2916023/87442184-97863c80-c5b9-11ea-8c51-21c450527298.png)

Try running `yarn upgrade @reach/router`

**Comment by blainekasten** [comment]
Closing this because it's not a bug with Gatsby itself, and it is indeed fixed with a new version of reach router. Feel free to keep responding here if you need further assistance! 💜

**Comment by blainekasten** [comment]
Cool! So this confirms my suspicion. This was fixed in @reach/router v1.3.0, try running `yarn upgrade @reach/router` and then verify you get a version greater than 1.3.0.

I'm going to close this as it's not an issue with Gatsby itself, and it's been fixed in @reach/router, feel free to keep discussing here if you need further assistance. 💜
